backends/ebpf/psa/ebpfPipeline.cpp

```diff
@@ -0,0 +1,607 @@
+                          "            .class_of_service = %s,\n"
+                          "            .egress_port = %s,\n"
+                          "            .packet_path = %s,\n"
+                          "            .instance = %s,\n"
+                          "            .parser_error = %s,\n"
+                          "        };",
+                          control->inputStandardMetadata->name.name,
+                          priorityVar.c_str(), ifindexVar.c_str(),
+                          packetPathVar.c_str(), pktInstanceVar.c_str(), errorVar.c_str());
+    builder->newline();
+    if (shouldEmitTimestamp()) {
+        builder->emitIndent();
+        builder->appendFormat("%s.egress_timestamp = %s",
+                              control->inputStandardMetadata->name.name,
+                              timestampVar.c_str());
+        builder->endOfStatement(true);
+    }
+    builder->emitIndent();
+    builder->appendFormat("if (%s.egress_port == PSA_PORT_RECIRCULATE) ",
+                          control->inputStandardMetadata->name.name);
+    builder->blockStart();
+    builder->emitIndent();
+    // To be conformant with psa.p4, where PSA_PORT_RECIRCULATE is constant
+    builder->appendFormat("%s.egress_port = P4C_PSA_PORT_RECIRCULATE",
+                          control->inputStandardMetadata->name.name);
+    builder->endOfStatement(true);
+    builder->blockEnd(true);
+}
+
+void EBPFEgressPipeline::emitPSAControlOutputMetadata(CodeBuilder *builder) {
+    builder->emitIndent();
+    builder->appendFormat("struct psa_egress_output_metadata_t %s = {\n",
+                          control->outputStandardMetadata->name.name);
+    builder->appendLine("       .clone = false,\n"
+                        "            .drop = false,\n"
+                        "        };");
+
+    builder->newline();
+}
+
+void EBPFEgressPipeline::emitCPUMAPLookup(CodeBuilder *builder) {
+    builder->emitIndent();
+    builder->target->emitTableLookup(builder, "hdr_md_cpumap", oneKey.c_str(), "hdrMd");
+    builder->endOfStatement(true);
+}
+
+void EBPFEgressPipeline::emit(CodeBuilder *builder) {
+    cstring msgStr, varStr;
+
+    builder->newline();
+    builder->target->emitCodeSection(builder, sectionName);
+    builder->emitIndent();
+    builder->target->emitMain(builder, functionName, model.CPacketName.str());
+    builder->spc();
+    builder->blockStart();
+
+    emitGlobalMetadataInitializer(builder);
+    emitLocalVariables(builder);
+    emitUserMetadataInstance(builder);
+    builder->newline();
+
+    emitHeaderInstances(builder);
+    builder->newline();
+
+    emitCPUMAPInitializers(builder);
+    builder->newline();
+    emitHeadersFromCPUMAP(builder);
+    builder->newline();
+    emitMetadataFromCPUMAP(builder);
+    builder->newline();
+
+    emitPSAControlOutputMetadata(builder);
+    emitPSAControlInputMetadata(builder);
+
+    msgStr = Util::printf_format("%s parser: parsing new packet, path=%%d, pkt_len=%%d",
```

Comment:
shouldn't this code be in the parser code generator?

Reply:
We need `lengthVar`, `sectionName` from `ebpfProgram` and `compilerGlobalMetadata` from `ebpfPipeline`. So, I suggest keeping it here.